File: packages/typescript/tests/currency-transaction-vectors.test.ts
/**
 * Currency transaction test vector validation
 *
 * Validates TypeScript implementation against reference test vectors from tessellation
 */

import { describe, expect, it } from '@jest/globals';
import * as fs from 'fs';
import * as path from 'path';
import {
  createCurrencyTransaction,
  verifyCurrencyTransaction,
  hashCurrencyTransaction,
  encodeCurrencyTransaction,
} from '../src/currency-transaction';
import { keyStore } from '@stardust-collective/dag4-keystore';
import type { CurrencyTransaction } from '../src/currency-types';
import { getAddress } from '../src/wallet';

// Load test vectors
const testVectorsPath = path.join(__dirname, '../../../shared/currency_transaction_vectors.json');
const testVectors = JSON.parse(fs.readFileSync(testVectorsPath, 'utf-8'));

describe('Currency Transaction Test Vectors', () => {
  const basic = testVectors.testVectors.basicTransaction;

  describe('Key Derivation', () => {
    it('derives correct public key from private key', () => {
      const publicKey = keyStore.getPublicKeyFromPrivate(basic.privateKeyHex);
      // dag4.js may return with or without 04 prefix depending on version
      const normalizedPublicKey = publicKey.startsWith('04') ? publicKey : '04' + publicKey;
      expect(normalizedPublicKey).toBe(basic.publicKeyHex);
    });

    it('derives correct address from public key', () => {
      const publicKey = keyStore.getPublicKeyFromPrivate(basic.privateKeyHex);
      const address = getAddress(publicKey);
      expect(address).toBe(basic.address);
    });
  });

  describe('Transaction Encoding', () => {
    it('encodes transaction correctly', async () => {
      // Create transaction with known salt
      const tx = await createCurrencyTransaction(
        {
          destination: basic.transaction.destination,
          amount: basic.transaction.amount / 1e8, // Convert to token units
          fee: basic.transaction.fee / 1e8,
        },
        basic.privateKeyHex,
        basic.transaction.parent
      );

      // Override salt for deterministic test
      (tx.value as any).salt = basic.transaction.salt.toString();

      const encoded = encodeCurrencyTransaction(tx);
      expect(encoded).toBe(basic.encodedString);
    });

    it('validates encoding format breakdown', async () => {
      const breakdown = testVectors.testVectors.encodingBreakdown;
      const components = breakdown.components;

      // Create transaction with known values
      const tx = await createCurrencyTransaction(
        {
          destination: components.destination.value,
          amount: parseInt(components.amountHex.value, 16) / 1e8,
          fee: parseInt(components.fee.value) / 1e8,
        },
        basic.privateKeyHex,
        {
          hash: components.parentHash.value,
          ordinal: parseInt(components.ordinal.value),
        }
      );

      // Override salt
      (tx.value as any).salt = parseInt(components.saltHex.value, 16).toString();

      const encoded = encodeCurrencyTransaction(tx);

      // Verify encoding format
      expect(encoded.startsWith('2')).toBe(true); // Version prefix
      expect(encoded).toContain(components.source.value);
      expect(encoded).toContain(components.destination.value);
      expect(encoded).toContain(components.amountHex.value);
      expect(encoded).toContain(components.parentHash.value);
    });
  });

  describe('Transaction Hashing', () => {
    it('produces correct transaction hash', async () => {
      // Create transaction with deterministic values
      const tx = await createCurrencyTransaction(
        {
          destination: basic.transaction.destination,
          amount: basic.transaction.amount / 1e8,
          fee: basic.transaction.fee / 1e8,
        },
        basic.privateKeyHex,
        basic.transaction.parent
      );

      // Override salt and proofs for exact match
      (tx.value as any).salt = basic.transaction.salt.toString();
      tx.proofs = [];

      const hash = await hashCurrencyTransaction(tx);
      expect(hash.value).toBe(basic.transactionHash);
    });
  });

  describe('Signature Verification', () => {
    it('verifies reference signature from Scala test vectors', async () => {
      // Scala signatures are normalized to low-S during verification
      const tx: CurrencyTransaction = {
        value: {
          ...basic.transaction,
          salt: basic.transaction.salt.toString(),
        },
        proofs: [
          {
            id: basic.signerId,
            signature: basic.signature,
          },
        ],
      };

      const result = await verifyCurrencyTransaction(tx);
      expect(result.isValid).toBe(true);
      expect(result.validProofs.length).toBe(1);
      expect(result.invalidProofs.length).toBe(0);
    });

    it('verifies multi-signature transaction from Scala test vectors', async () => {
      // Scala signatures are normalized to low-S during verification
      const multiSig = testVectors.testVectors.multiSignature;
      expect(multiSig.transactionHash).toBe(basic.transactionHash);

      const tx: CurrencyTransaction = {
        value: {
          ...basic.transaction,
          salt: basic.transaction.salt.toString(),
        },
        proofs: multiSig.proofs,
      };

      const result = await verifyCurrencyTransaction(tx);
      expect(result.isValid).toBe(true);
      expect(result.validProofs.length).toBe(2);
      expect(result.invalidProofs.length).toBe(0);

      // Verify both proofs are marked as valid in test vectors
      multiSig.proofs.forEach((proof: any) => {
        expect(proof.valid).toBe(true);
      });
    });
  });

  describe('Transaction Chaining', () => {
    it('validates transaction chain parent references', () => {
      const chain = testVectors.testVectors.transactionChaining.transactions;

      // Verify first transaction parent
      expect(chain[0].parentHash).toBe('aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa');
      expect(chain[0].parentOrdinal).toBe(5);
      expect(chain[0].ordinal).toBe(6);

      // Verify second transaction chains to first
      expect(chain[1].parentHash).toBe(chain[0].hash);
      expect(chain[1].parentOrdinal).toBe(chain[0].ordinal);
      expect(chain[1].ordinal).toBe(7);

      // Verify third transaction chains to second
      expect(chain[2].parentHash).toBe(chain[1].hash);
      expect(chain[2].parentOrdinal).toBe(chain[1].ordinal);
      expect(chain[2].ordinal).toBe(8);
    });
  });

  describe('Edge Cases', () => {
    it('validates minimum amount transaction', () => {
      const minAmount = testVectors.testVectors.edgeCases.minAmount;
      expect(minAmount.amount).toBe(1);
      expect(minAmount.hash).toBeTruthy();
      expect(minAmount.signature).toBeTruthy();
    });

    it('validates maximum amount transaction', () => {
      // Note: JavaScript's MAX_SAFE_INTEGER is 2^53-1 (9007199254740991)
      // Long.MAX_VALUE is 2^63-1 (9223372036854775807)
      // JavaScript cannot accurately represent Long.MAX_VALUE as a number
      // However, we can still validate that the hash and signature exist
      const maxAmount = testVectors.testVectors.edgeCases.maxAmount;
      expect(maxAmount.hash).toBeTruthy();
      expect(maxAmount.signature).toBeTruthy();
      // Note: We don't validate maxAmount.amount here due to JavaScript number precision limits
    });

    it('validates transaction with non-zero fee', () => {
      const withFee = testVectors.testVectors.edgeCases.withFee;
      expect(withFee.amount).toBe(10000000000);
      expect(withFee.fee).toBe(100000);
      expect(withFee.hash).toBeTruthy();
      expect(withFee.signature).toBeTruthy();
    });
  });

  describe('Kryo Serialization', () => {
    it('validates Kryo setReferences=false format (v2 transactions)', () => {
      const params = testVectors.cryptoParams;
      expect(params.kryoSetReferences).toBe(false);
    });

    it('validates Kryo header without reference flag (v2 format)', () => {
      const kryoHex = basic.kryoBytesHex;
      // Should start with 0x03 (string type) followed by length, no 0x01 reference flag for v2
      expect(kryoHex.startsWith('03')).toBe(true);
      expect(kryoHex.startsWith('0301')).toBe(false); // No reference flag for v2
    });
  });
});

File: packages/go/currency_transaction_vectors_test.go
package constellation

import (
	"encoding/hex"
	"encoding/json"
	"os"
	"path/filepath"
	"strings"
	"testing"

	"github.com/btcsuite/btcd/btcec/v2"
)

// testCurrencyTransactionValue is a wrapper for test vectors that can handle salt as number or string
type testCurrencyTransactionValue struct {
	Source      string               `json:"source"`
	Destination string               `json:"destination"`
	Amount      int64                `json:"amount"`
	Fee         int64                `json:"fee"`
	Parent      TransactionReference `json:"parent"`
	Salt        json.Number          `json:"salt"` // Can be number or string in JSON
}

// toCurrencyTransactionValue converts test value to runtime value
func (t *testCurrencyTransactionValue) toCurrencyTransactionValue() CurrencyTransactionValue {
	return CurrencyTransactionValue{
		Source:      t.Source,
		Destination: t.Destination,
		Amount:      t.Amount,
		Fee:         t.Fee,
		Parent:      t.Parent,
		Salt:        string(t.Salt),
	}
}

// TestVectors represents the structure of the test vectors JSON
type TestVectors struct {
	CryptoParams struct {
		KryoSetReferences bool `json:"kryoSetReferences"`
	} `json:"cryptoParams"`
	TestVectors struct {
		BasicTransaction struct {
			PrivateKeyHex   string                        `json:"privateKeyHex"`
			PublicKeyHex    string                        `json:"publicKeyHex"`
			Address         string                        `json:"address"`
			Transaction     testCurrencyTransactionValue  `json:"transaction"`
			EncodedString   string                        `json:"encodedString"`
			KryoBytesHex    string                        `json:"kryoBytesHex"`
			TransactionHash string                        `json:"transactionHash"`
			Signature       string                        `json:"signature"`
			SignerID        string                        `json:"signerId"`
		} `json:"basicTransaction"`
		EncodingBreakdown struct {
			Components struct {
				VersionPrefix string `json:"versionPrefix"`
				Source        struct {
					Length int    `json:"length"`
					Value  string `json:"value"`
				} `json:"source"`
				Destination struct {
					Length int    `json:"length"`
					Value  string `json:"value"`
				} `json:"destination"`
				AmountHex struct {
					Length int    `json:"length"`
					Value  string `json:"value"`
				} `json:"amountHex"`
				ParentHash struct {
					Length int    `json:"length"`
					Value  string `json:"value"`
				} `json:"parentHash"`
			} `json:"components"`
			FullEncoded string `json:"fullEncoded"`
		} `json:"encodingBreakdown"`
		MultiSignature struct {
			TransactionHash string            `json:"transactionHash"`
			Proofs          []SignatureProof  `json:"proofs"`
		} `json:"multiSignature"`
		TransactionChaining struct {
			Transactions []struct {
				Index        int    `json:"index"`
				Hash         string `json:"hash"`
				Ordinal      int    `json:"ordinal"`
				ParentHash   string `json:"parentHash"`
				ParentOrdinal int   `json:"parentOrdinal"`
			} `json:"transactions"`
		} `json:"transactionChaining"`
		EdgeCases struct {
			MinAmount struct {
				Amount    int64  `json:"amount"`
				Hash      string `json:"hash"`
				Signature string `json:"signature"`
			} `json:"minAmount"`
			MaxAmount struct {
				Amount    int64  `json:"amount"`
				Hash      string `json:"hash"`
				Signature string `json:"signature"`
			} `json:"maxAmount"`
			WithFee struct {
				Amount    int64  `json:"amount"`
				Fee       int64  `json:"fee"`
				Hash      string `json:"hash"`
				Signature string `json:"signature"`
			} `json:"withFee"`
		} `json:"edgeCases"`
	} `json:"testVectors"`
}

func loadCurrencyTestVectors(t *testing.T) *TestVectors {
	vectorsPath := filepath.Join("..", "..", "shared", "currency_transaction_vectors.json")
	data, err := os.ReadFile(vectorsPath)
	if err != nil {
		t.Fatalf("Failed to read test vectors: %v", err)
	}

	var vectors TestVectors
	if err := json.Unmarshal(data, &vectors); err != nil {
		t.Fatalf("Failed to parse test vectors: %v", err)
	}

	return &vectors
}

func TestKeyDerivation(t *testing.T) {
	vectors := loadCurrencyTestVectors(t)
	basic := vectors.TestVectors.BasicTransaction

	t.Run("derives correct public key", func(t *testing.T) {
		privateKeyBytes, err := hex.DecodeString(basic.PrivateKeyHex)
		if err != nil {
			t.Fatalf("Failed to decode private key: %v", err)
		}

		privateKey, _ := btcec.PrivKeyFromBytes(privateKeyBytes)
		publicKeyHex := hex.EncodeToString(privateKey.PubKey().SerializeUncompressed())

		if publicKeyHex != basic.PublicKeyHex {
			t.Errorf("Public key mismatch: got %s, want %s", publicKeyHex, basic.PublicKeyHex)
		}
	})

	t.Run("derives correct address", func(t *testing.T) {
		privateKeyBytes, err := hex.DecodeString(basic.PrivateKeyHex)
		if err != nil {
			t.Fatalf("Failed to decode private key: %v", err)
		}

		privateKey, _ := btcec.PrivKeyFromBytes(privateKeyBytes)
		publicKeyHex := hex.EncodeToString(privateKey.PubKey().SerializeUncompressed())
		address := GetAddress(publicKeyHex)

		if address != basic.Address {
			t.Errorf("Address mismatch: got %s, want %s", address, basic.Address)
		}
	})
}

func TestTransactionEncoding(t *testing.T) {
	vectors := loadCurrencyTestVectors(t)
	basic := vectors.TestVectors.BasicTransaction
	txValue := basic.Transaction.toCurrencyTransactionValue()

	t.Run("encodes transaction correctly", func(t *testing.T) {
		// Create transaction with known values
		tx, err := CreateCurrencyTransaction(
			TransferParams{
				Destination: txValue.Destination,
				Amount:      float64(txValue.Amount) / 1e8,
				Fee:         float64(txValue.Fee) / 1e8,
			},
			basic.PrivateKeyHex,
			txValue.Parent,
		)
		if err != nil {
			t.Fatalf("Failed to create transaction: %v", err)
		}

		// Override salt for deterministic test
		tx.Value.Salt = txValue.Salt
		tx.Proofs = []SignatureProof{}

		encoded := encodeTransaction(tx)
		if encoded != basic.EncodedString {
			t.Errorf("Encoding mismatch:\ngot:  %s\nwant: %s", encoded, basic.EncodedString)
		}
	})

	t.Run("validates encoding breakdown", func(t *testing.T) {
		breakdown := vectors.TestVectors.EncodingBreakdown
		components := breakdown.Components

		// Verify version prefix
		if components.VersionPrefix != "2" {
			t.Errorf("Version prefix mismatch: got %s, want 2", components.VersionPrefix)
		}

		// Verify components are present in full encoding
		fullEncoded := breakdown.FullEncoded
		if fullEncoded[0] != '2' {
			t.Error("Encoding should start with version prefix '2'")
		}
	})
}

func TestCurrencyTransactionVectorsHashing(t *testing.T) {
	vectors := loadCurrencyTestVectors(t)
	basic := vectors.TestVectors.BasicTransaction
	txValue := basic.Transaction.toCurrencyTransactionValue()

	t.Run("produces correct transaction hash", func(t *testing.T) {
		// Create transaction with deterministic values
		tx, err := CreateCurrencyTransaction(
			TransferParams{
				Destination: txValue.Destination,
				Amount:      float64(txValue.Amount) / 1e8,
				Fee:         float64(txValue.Fee) / 1e8,
			},
			basic.PrivateKeyHex,
			txValue.Parent,
		)
		if err != nil {
			t.Fatalf("Failed to create transaction: %v", err)
		}

		// Override salt for exact match
		tx.Value.Salt = txValue.Salt
		tx.Proofs = []SignatureProof{}

		hash := HashCurrencyTransaction(tx)
		if hash.Value != basic.TransactionHash {
			t.Errorf("Hash mismatch:\ngot:  %s\nwant: %s", hash.Value, basic.TransactionHash)
		}
	})
}

func TestSignatureVerification(t *testing.T) {
	vectors := loadCurrencyTestVectors(t)
	basic := vectors.TestVectors.BasicTransaction
	txValue := basic.Transaction.toCurrencyTransactionValue()

	t.Run("verifies reference signature", func(t *testing.T) {
		// Reconstruct transaction from test vector
		tx := &CurrencyTransaction{
			Value: txValue,
			Proofs: []SignatureProof{
				{
					ID:        basic.SignerID,
					Signature: basic.Signature,
				},
			},
		}

		result := VerifyCurrencyTransaction(tx)
		if !result.IsValid {
			t.Error("Transaction should be valid")
		}
		if len(result.ValidProofs) != 1 {
			t.Errorf("Expected 1 valid proof, got %d", len(result.ValidProofs))
		}
		if len(result.InvalidProofs) != 0 {
			t.Errorf("Expected 0 invalid proofs, got %d", len(result.InvalidProofs))
		}
	})

	t.Run("verifies multi-signature transaction", func(t *testing.T) {
		multiSig := vectors.TestVectors.MultiSignature

		tx := &CurrencyTransaction{
			Value:  txValue,
			Proofs: multiSig.Proofs,
		}

		result := VerifyCurrencyTransaction(tx)
		if !result.IsValid {
			t.Error("Multi-sig transaction should be valid")
		}
		if len(result.ValidProofs) != 2 {
			t.Errorf("Expected 2 valid proofs, got %d", len(result.ValidProofs))
		}
		if len(result.InvalidProofs) != 0 {
			t.Errorf("Expected 0 invalid proofs, got %d", len(result.InvalidProofs))
		}
	})
}

func TestTransactionChaining(t *testing.T) {
	vectors := loadCurrencyTestVectors(t)
	chain := vectors.TestVectors.TransactionChaining.Transactions

	t.Run("validates chain parent references", func(t *testing.T) {
		// Verify first transaction parent
		if chain[0].ParentHash != "aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa" {
			t.Error("First transaction parent hash mismatch")
		}
		if chain[0].ParentOrdinal != 5 {
			t.Errorf("First transaction parent ordinal: got %d, want 5", chain[0].ParentOrdinal)
		}
		if chain[0].Ordinal != 6 {
			t.Errorf("First transaction ordinal: got %d, want 6", chain[0].Ordinal)
		}

		// Verify second transaction chains to first
		if chain[1].ParentHash != chain[0].Hash {
			t.Error("Second transaction should chain to first")
		}
		if chain[1].ParentOrdinal != chain[0].Ordinal {
			t.Error("Second transaction parent ordinal should match first ordinal")
		}
		if chain[1].Ordinal != 7 {
			t.Errorf("Second transaction ordinal: got %d, want 7", chain[1].Ordinal)
		}

		// Verify third transaction chains to second
		if chain[2].ParentHash != chain[1].Hash {
			t.Error("Third transaction should chain to second")
		}
		if chain[2].ParentOrdinal != chain[1].Ordinal {
			t.Error("Third transaction parent ordinal should match second ordinal")
		}
		if chain[2].Ordinal != 8 {
			t.Errorf("Third transaction ordinal: got %d, want 8", chain[2].Ordinal)
		}
	})
}

func TestEdgeCases(t *testing.T) {
	vectors := loadCurrencyTestVectors(t)
	edgeCases := vectors.TestVectors.EdgeCases

	t.Run("validates minimum amount", func(t *testing.T) {
		minAmount := edgeCases.MinAmount
		if minAmount.Amount != 1 {
			t.Errorf("Min amount: got %d, want 1", minAmount.Amount)
		}
		if minAmount.Hash == "" {
			t.Error("Min amount hash should not be empty")
		}
		if minAmount.Signature == "" {
			t.Error("Min amount signature should not be empty")
		}
	})

	t.Run("validates maximum amount", func(t *testing.T) {
		maxAmount := edgeCases.MaxAmount
		if maxAmount.Amount != 9223372036854775807 {
			t.Errorf("Max amount: got %d, want 9223372036854775807", maxAmount.Amount)
		}
		if maxAmount.Hash == "" {
			t.Error("Max amount hash should not be empty")
		}
		if maxAmount.Signature == "" {
			t.Error("Max amount signature should not be empty")
		}
	})

	t.Run("validates non-zero fee", func(t *testing.T) {
		withFee := edgeCases.WithFee
		if withFee.Amount != 10000000000 {
			t.Errorf("Fee amount: got %d, want 10000000000", withFee.Amount)
		}
		if withFee.Fee != 100000 {
			t.Errorf("Fee: got %d, want 100000", withFee.Fee)
		}
		if withFee.Hash == "" {
			t.Error("Fee hash should not be empty")
		}
		if withFee.Signature == "" {
			t.Error("Fee signature should not be empty")
		}
	})
}

func TestKryoSerialization(t *testing.T) {
	vectors := loadCurrencyTestVectors(t)

	t.Run("validates setReferences flag (v2 format)", func(t *testing.T) {
		if vectors.CryptoParams.KryoSetReferences {
			t.Error("Kryo setReferences should be false for v2 transactions")
		}
	})

	t.Run("validates Kryo header without reference flag (v2 format)", func(t *testing.T) {
		basic := vectors.TestVectors.BasicTransaction
		kryoHex := basic.KryoBytesHex
		// Should start with 0x03 (string type) followed by length, no 0x01 reference flag for v2
		if !strings.HasPrefix(kryoHex, "03") {
			t.Errorf("Kryo header should start with 03, got %s", kryoHex[:2])
		}
		if strings.HasPrefix(kryoHex, "0301") {
			t.Error("Kryo header should NOT have reference flag (0301) for v2 transactions")
		}
	})
}
